README.rst
========
img2cmap
========

Usage
=====

Basic
-----

**Create colormaps from images in three lines of code!**

| First, ``ImageConverter`` class converts images to arrays of RGB values.
| Then, ``generate_cmap`` creates a matplotlib `ListedColormap <https://matplotlib.org/stable/api/_as_gen/matplotlib.colors.ListedColormap.html#matplotlib-colors-listedcolormap>`_.

.. code-block:: python3

    from img2cmap import ImageConverter

    # Can be a local file or URL
    converter = ImageConverter("tests/images/south_beach_sunset.jpg")
    cmap = converter.generate_cmap(n_colors=5, palette_name="south_beach_sunset", random_state=42)


Plot an image and a colorbar side by side.

.. code-block:: python3

    import matplotlib.pyplot as plt
    from mpl_toolkits.axes_grid1 import make_axes_locatable

    fig, ax = plt.subplots(figsize=(7, 5))

    ax.axis("off")
    img = plt.imread("tests/images/south_beach_sunset.jpg")
    im = ax.imshow(img, cmap=cmap)

    divider = make_axes_locatable(ax)
    cax = divider.append_axes("right", size="10%", pad=0.05)

    cb = fig.colorbar(im, cax=cax, orientation="vertical", label=cmap.name)
    cb.set_ticks([])

.. image:: images/colorbar.png
    :align: center

Now, use the colormap in your plots!

.. code-block:: python3

    import matplotlib.pyplot as plt

    colors = cmap.colors

    with plt.style.context("dark_background"):
        for i, color in enumerate(colors):
            plt.plot(range(10), [_+i+1 for _ in range(10)], color=color, linewidth=4)

.. image:: images/img2cmap_demo.png
    :align: center


Advanced
--------

remove_transparency
^^^^^^^^^^^^^^^^^^^

In an image of the Los Angeles Lakers logo, the background is transparent. These pixels
contribute to noise when generating the colors. Setting ``remove_transparency`` to ``True``
will remove transparent pixels. Here's a comparison of the colormaps generated by the same image,
without and with transparency removed.

Make two ImageConverter objects:

.. code-block:: python3

    from img2cmap import ImageConverter

    image_url = "https://loodibee.com/wp-content/uploads/nba-los-angeles-lakers-logo.png"

    # Create two ImageConverters, one with transparency removed and one without
    converter_with_transparent = ImageConverter(image_url, remove_transparent=False)
    converter_no_transparent = ImageConverter(image_url, remove_transparent=True)

    cmap_with_transparent = converter_with_transparent.generate_cmap(
        n_colors=3, palette_name="with_transparent", random_state=42
    )
    cmap_no_transparent = converter_no_transparent.generate_cmap(
        n_colors=3, palette_name="no_transparent", random_state=42
    )

Plot both colormaps with the image:


.. code-block:: python3

    import matplotlib.pyplot as plt
    from mpl_toolkits.axes_grid1 import make_axes_locatable

    for cmap in [cmap_with_transparent, cmap_no_transparent]:
        fig, ax = plt.subplots(figsize=(7, 5))

        ax.axis("off")
        img = converter_no_transparent.image
        im = ax.imshow(img, cmap=cmap)

        divider = make_axes_locatable(ax)
        cax = divider.append_axes("right", size="10%", pad=0.05)

        cb = fig.colorbar(im, cax=cax, orientation="vertical", label=cmap.name)
        cb.set_ticks([])


.. image:: images/lakers_with_transparent.png
    :align: center

.. image:: images/lakers_no_transparent.png
    :align: center

Notice, only after removing the transparent pixels, does the classic purple and gold show in the colormap.


generate_optimal_cmap
^^^^^^^^^^^^^^^^^^^^^

You can extract the optimal number of colors from the image using the ``generate_optimal_cmap`` method.
Under the hood this performs the `elbow method <https://en.wikipedia.org/wiki/Elbow_method_(clustering)>`
to determine the optimal number of clusters based on the sum of the squared distances between each pixel
and it's cluster center.


.. code-block:: python3

    cmaps, best_n_colors, ssd = converter.generate_optimal_cmap(max_colors=10, random_state=42)

    best_cmap = cmaps[best_n_colors]


Installation
============

::

    pip install img2cmap

You can also install the in-development version with::

    pip install https://github.com/arvkevi/img2cmap/archive/main.zip


Documentation
=============


https://img2cmap.readthedocs.io/


Status
======


.. start-badges

.. list-table::
    :stub-columns: 1

    * - docs
      - |docs|
    * - tests
      - | |github-actions|
        | |codecov|
    * - package
      - | |version| |wheel| |supported-versions| |supported-implementations|
.. |docs| image:: https://readthedocs.org/projects/img2cmap/badge/?style=flat
    :target: https://img2cmap.readthedocs.io/
    :alt: Documentation Status

.. |github-actions| image:: https://github.com/arvkevi/img2cmap/actions/workflows/github-actions.yml/badge.svg
    :alt: GitHub Actions Build Status
    :target: https://github.com/arvkevi/img2cmap/actions

.. |codecov| image:: https://codecov.io/gh/arvkevi/img2cmap/branch/main/graphs/badge.svg?branch=main
    :alt: Coverage Status
    :target: https://codecov.io/github/arvkevi/img2cmap

.. |version| image:: https://img.shields.io/pypi/v/img2cmap.svg
    :alt: PyPI Package latest release
    :target: https://pypi.org/project/img2cmap

.. |wheel| image:: https://img.shields.io/pypi/wheel/img2cmap.svg
    :alt: PyPI Wheel
    :target: https://pypi.org/project/img2cmap

.. |supported-versions| image:: https://img.shields.io/pypi/pyversions/img2cmap.svg
    :alt: Supported versions
    :target: https://pypi.org/project/img2cmap

.. |supported-implementations| image:: https://img.shields.io/pypi/implementation/img2cmap.svg
    :alt: Supported implementations
    :target: https://pypi.org/project/img2cmap



.. end-badges


Development
===========

To run all the tests run::

    tox

Note, to combine the coverage data from all the tox environments run:

.. list-table::
    :widths: 10 90
    :stub-columns: 1

    - - Windows
      - ::

            set PYTEST_ADDOPTS=--cov-append
            tox

    - - Other
      - ::

            PYTEST_ADDOPTS=--cov-append tox


CHANGELOG.rst

Changelog
=========

0.0.0 (2022-04-30)
------------------

* First release on PyPI.


tests/test_img2cmap.py
from pathlib import Path

# import matplotlib as mpl
import pytest

from img2cmap import ImageConverter

THIS_DIR = Path(__file__).parent

images = list(THIS_DIR.joinpath("images").iterdir())


@pytest.mark.parametrize("test_image_input", images)
def test_generate_cmap_1(test_image_input):
    imageconverter = ImageConverter(test_image_input)
    cmap = imageconverter.generate_cmap(4, "miami", 42)
    assert cmap.name == "miami"


@pytest.mark.parametrize("test_image_input", images)
def test_generate_cmap_2(test_image_input):
    imageconverter = ImageConverter(test_image_input)
    cmap = imageconverter.generate_cmap(4, "miami", 42)
    assert cmap.N == 4


@pytest.mark.parametrize("test_image_input", images)
def test_generate_cmap_3(test_image_input):
    cmap_default_name = test_image_input.stem
    imageconverter = ImageConverter(test_image_input)
    cmap = imageconverter.generate_cmap(4, None, 42)
    assert cmap.name == cmap_default_name


@pytest.mark.parametrize("test_image_input", images)
def test_generate_cmap_4(test_image_input):
    with pytest.raises(ValueError):
        imageconverter = ImageConverter(test_image_input)
        imageconverter.generate_cmap(-100, "miami", 42)


@pytest.mark.parametrize("test_image_input", images)
def test_cmap_color_dimension(test_image_input):
    imageconverter = ImageConverter(test_image_input)
    cmap = imageconverter.generate_cmap(4, "Miami Yeaaaa", 42)
    assert cmap.colors.shape[1] == 3


# TODO: Mock these!
# def test_url():
#     with open(THIS_DIR.joinpath("urls/nba-logos.txt"), "r") as f:
#         for line in f:
#             if "miami" in line:
#                 url = line.strip()
#                 break
#     converter = ImageConverter(url)
#     cmap = converter.generate_cmap(2, "miami", 42)
#     assert cmap.name == "miami"


# @pytest.mark.parametrize("test_remove_transparent", [True, False])
# def test_remove_transparent(test_remove_transparent):
#     """This image should not have any black pixels."""
#     with open(THIS_DIR.joinpath("urls/nba-logos.txt"), "r") as f:
#         for line in f:
#             if "atlanta" in line:
#                 url = line.strip()
#     converter = ImageConverter(url, remove_transparent=test_remove_transparent)
#     cmap = converter.generate_cmap(3, 42)
#     hex_codes = [mpl.colors.rgb2hex(c) for c in cmap.colors]
#     black_not_in_colors = not any(["#000000" in c for c in hex_codes])
#     assert black_not_in_colors == test_remove_transparent


def test_generate_optimal():
    imageconverter = ImageConverter(THIS_DIR.joinpath("images/south_beach_sunset.jpg"))
    _, best_n_colors, _ = imageconverter.generate_optimal_cmap(random_state=42)
    assert best_n_colors == 5


src/img2cmap/convert.py
from pathlib import Path
from urllib.error import HTTPError
from urllib.error import URLError
from urllib.request import urlopen

import matplotlib as mpl
import numpy as np
from kneed import KneeLocator
from PIL import Image
from sklearn.cluster import MiniBatchKMeans


class ImageConverter:
    """Converts an image to numpy array of RGB values.

    Args:
        image_path str: The path to the image. Can be a local file or a URL.
        remove_transparent bool: If True, will not consider any transparent pixels. Defaults to False.
        downsample bool: If True, will downsample the image to 512x512. Defaults to True.

    Attributes:
        image_path (str): The path to the image. Can be a local file or a URL.
        image (PIL.Image): The image object.
        pixels (numpy.ndarray): A numpy array of RGB values.
        kmeans (sklearn.cluster.MiniBatchKMeans): A kmeans model.
    """

    def __init__(self, image_path, remove_transparent=False, downsample=True):
        self.image_path = image_path
        # try to open the image
        try:
            self.image = Image.open(self.image_path)
        except FileNotFoundError:
            try:
                self.image = Image.open(urlopen(self.image_path))
            except (URLError, HTTPError, FileNotFoundError) as error:
                raise URLError(f"Could not open {self.image_path} {error}") from error

        # downsample the image with antialiasing, inplace
        if downsample:
            self.image.thumbnail((512, 512), Image.LANCZOS)
        # convert the image to a numpy array
        self.image = self.image.convert("RGBA")
        self.pixels = np.array(self.image.getdata())
        if remove_transparent:
            self.pixels = self.pixels[self.pixels[:, 3] != 0]
        self.pixels = self.pixels[:, :3]
        self.kmeans = None

    def generate_cmap(self, n_colors=4, palette_name=None, random_state=None):
        """Generates a matplotlib ListedColormap from an image.

        Args:
            n_colors (int, optional): The number of colors in the ListedColormap. Defaults to 4.
            palette_name (str, optional): A name for your created palette. If None, defaults to the image name.
                Defaults to None.
            random_state (int, optional): A random seed for reproducing ListedColormaps.
                The k-means algorithm has a random initialization step and doesn't always converge on the same
                solution because of this. If None will be a different seed each time this method is called.
                Defaults to None.

        Returns:
            matplotlib.colors.ListedColormap: A matplotlib ListedColormap object.
        """
        # create a kmeans model
        self.kmeans = MiniBatchKMeans(n_clusters=n_colors, random_state=random_state)
        # fit the model to the pixels
        self.kmeans.fit(self.pixels)
        # get the cluster centers
        centroids = self.kmeans.cluster_centers_ / 255
        # return the palette
        if palette_name is None:
            palette_name = Path(self.image_path).stem

        cmap = mpl.colors.ListedColormap(centroids, name=palette_name)

        # Handle 4 dimension RGBA colors
        cmap.colors = cmap.colors[:, :3]
        return cmap

    def generate_optimal_cmap(self, max_colors=10, palette_name=None, random_state=None):
        """Generates an optimal matplotlib ListedColormap from an image by finding the optimal number of clusters using the elbow method.

        Useage:
            >>> img = ImageConverter("path/to/image.png")
            >>> cmaps, best_n_colors, ssd = img.generate_optimal_cmap()
            >>> # The optimal colormap
            >>> cmaps[best_n_colors]


        Args:
            max_colors (int, optional): _description_. Defaults to 10.
            palette_name (_type_, optional): _description_. Defaults to None.
            random_state (_type_, optional): _description_. Defaults to None.
            remove_background (_type_, optional): _description_. Defaults to None.

        Returns:

            dict: A dictionary of matplotlib ListedColormap objects.
            Keys are the number of colors (clusters). Values are ListedColormap objects.
            int: The optimal number of colors.
            dict: A dictionary of the sum of square distances from each point to the cluster center.
            Keys are the number of colors (clusters) and values are the SSD value.
        """
        ssd = dict()
        cmaps = dict()
        for n_colors in range(2, max_colors + 1):
            cmap = self.generate_cmap(n_colors=n_colors, palette_name=palette_name, random_state=random_state)
            cmaps[n_colors] = cmap
            ssd[n_colors] = self.kmeans.inertia_

        best_n_colors = KneeLocator(list(ssd.keys()), list(ssd.values()), curve="convex", direction="decreasing").knee

        return cmaps, best_n_colors, ssd


streamlit/app.py
import warnings
from io import BytesIO

import matplotlib as mpl
import matplotlib.patches as patches
import matplotlib.pyplot as plt
import numpy as np
from mpl_toolkits.axes_grid1 import make_axes_locatable

import streamlit as st
from img2cmap import ImageConverter

warnings.filterwarnings("ignore")
st.set_option("deprecation.showfileUploaderEncoding", False)

st.set_page_config(
    page_title="img2cmap web",
    layout="wide",
)

st.title("Convert images to a colormap")
st.markdown(
    """
    This app converts images to colormaps using the Python
    library [img2cmap](https://github.com/arvkevi/img2cmap).
    Try your own image on the left. **Scroll down to generate an optimal colormap.**
    """
)

st.sidebar.markdown("### Image settings")
file_or_url = st.sidebar.radio("Upload an image file or paste an image URL", ("file", "url"))

if file_or_url == "file":
    user_image = st.sidebar.file_uploader("Upload an image file")
    if user_image is not None:
        user_image = BytesIO(user_image.getvalue())
elif file_or_url == "url":
    user_image = st.sidebar.text_input("Paste an image URL", "https://static1.bigstockphoto.com/3/2/3/large1500/323952496.jpg")
else:
    st.warning("Please select an option")

# default image to use
if user_image is None:
    user_image = "https://raw.githubusercontent.com/arvkevi/img2cmap/main/tests/images/south_beach_sunset.jpg"

# user settings
st.sidebar.markdown("### User settings")
n_colors = st.sidebar.number_input(
    "Number of colors", min_value=2, max_value=20, value=5, help="The number of colors to return in the colormap"
)
n_colors = int(n_colors)
remove_transparent = st.sidebar.checkbox(
    "Remove transparency", False, help="If checked, remove transparent pixels from the image before clustering."
)
random_state = st.sidebar.number_input("Random state", value=42, help="Random state for reproducibility")
random_state = int(random_state)


@st.cache(allow_output_mutation=True)
def get_image_converter(user_image, remove_transparent):
    converter = ImageConverter(user_image, remove_transparent=remove_transparent)
    return converter


converter = get_image_converter(user_image, remove_transparent)

with st.spinner("Generating colormap..."):
    cmap = converter.generate_cmap(n_colors=n_colors, palette_name="", random_state=random_state)

# plot the image and colorbar
fig1, ax1 = plt.subplots(figsize=(8, 8))

ax1.axis("off")
img = converter.image
im = ax1.imshow(img, cmap=cmap)

divider = make_axes_locatable(ax1)
cax = divider.append_axes("right", size="10%", pad=0.05)

cb = fig1.colorbar(im, cax=cax, orientation="vertical", label=cmap.name)
cb.set_ticks([])
st.pyplot(fig1)

colors1 = [mpl.colors.rgb2hex(c) for c in cmap.colors]
st.text("Hex Codes (click to copy on far right)")
st.code(colors1)


st.header("Detect optimal number of colors")
optimize = st.button("Optimize")
if optimize:
    with st.spinner("Optimizing... (this takes about a minute)"):
        cmaps, best_n_colors, ssd = converter.generate_optimal_cmap(max_colors=20, palette_name="", random_state=random_state)

    figopt, ax = plt.subplots(figsize=(7, 5))

    ymax = 21
    xmax = 20
    ax.set_ylim(2, ymax)
    ax.set_xlim(0, 20)

    # i will be y axis
    for y, cmap_ in cmaps.items():
        colors = sorted([mpl.colors.rgb2hex(c) for c in cmap_.colors])
        intervals, width = np.linspace(0, xmax, len(colors) + 1, retstep=True)
        # j will be x axis
        for j, color in enumerate(colors):
            rect = patches.Rectangle((intervals[j], y), width, 1, facecolor=color)
            ax.add_patch(rect)

    ax.set_yticks(np.arange(2, ymax) + 0.5)
    ax.set_yticklabels(np.arange(2, ymax))
    ax.set_ylabel("Number of colors")
    ax.set_xticks([])

    # best
    rect = patches.Rectangle((0, best_n_colors), ymax, 1, linewidth=1, facecolor="none", edgecolor="black", linestyle="--")
    ax.add_patch(rect)

    # minus 2, one for starting at 2 and one for 0-indexing
    ax.get_yticklabels()[best_n_colors - 2].set_color("red")
    st.pyplot(figopt)
    st.metric("Optimal number of colors", best_n_colors)
    st.text("Hex Codes of optimal colormap (click to copy on far right)")
    st.code(sorted([mpl.colors.rgb2hex(c) for c in cmaps[best_n_colors].colors]))

    st.text("Sum of squared distances by number of colors:")
    st.write(ssd)
